File: contracts/api/clients/jakarta/src/main/java/app/epistola/client/jakarta/error/ProblemDetailParser.java
// SPDX-FileCopyrightText: Epistola Nederland B.V.
//
// SPDX-License-Identifier: EUPL-1.2

package app.epistola.client.jakarta.error;

import app.epistola.client.jakarta.EpistolaJson;
import app.epistola.client.jakarta.model.DataModelValidationError;
import app.epistola.client.jakarta.model.ProblemDetail;
import app.epistola.client.jakarta.model.ValidationError;
import jakarta.json.Json;
import jakarta.json.JsonArray;
import jakarta.json.JsonObject;
import jakarta.json.JsonReader;
import jakarta.json.JsonValue;
import jakarta.ws.rs.core.MediaType;
import java.io.StringReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Reads an {@code application/problem+json} body into the pieces
 * {@link ProblemDetailException} carries.
 *
 * <p>The generated {@code ProblemDetail}, {@code ValidationProblemDetail} and
 * {@code DataModelValidationProblemDetail} are independent classes, so the base problem and each
 * extension member are read separately: the body is parsed once with JSON-P and the extension
 * members are then bound with JSON-B. Members the contract does not model are ignored, as RFC 9457
 * requires.
 */
public final class ProblemDetailParser {

    /** The RFC 9457 problem media type, {@code application/problem+json}. */
    public static final MediaType PROBLEM_JSON = new MediaType("application", "problem+json");

    /** A parsed problem body: the base problem plus each extension member the contract models. */
    public static final class ParsedProblem {

        private final ProblemDetail problem;
        private final List<ValidationError> errors;
        private final Map<String, List<DataModelValidationError>> validationErrors;

        ParsedProblem(
                ProblemDetail problem,
                List<ValidationError> errors,
                Map<String, List<DataModelValidationError>> validationErrors) {
            this.problem = problem;
            this.errors = errors;
            this.validationErrors = validationErrors;
        }

        public ProblemDetail problem() {
            return problem;
        }

        /** Field-level failures from a {@code ValidationProblemDetail}; empty when absent. */
        public List<ValidationError> errors() {
            return errors;
        }

        /** Per-example failures from a {@code DataModelValidationProblemDetail}; empty when absent. */
        public Map<String, List<DataModelValidationError>> validationErrors() {
            return validationErrors;
        }
    }

    /**
     * True when {@code mediaType} is {@code application/problem+json} (or a subtype of it, such as
     * one carrying a charset parameter). A {@code null} media type is not a problem body.
     */
    public static boolean isProblemJson(MediaType mediaType) {
        return mediaType != null
                && "application".equalsIgnoreCase(mediaType.getType())
                && "problem+json".equalsIgnoreCase(mediaType.getSubtype());
    }

    /**
     * Parses {@code body} into a {@link ParsedProblem}, or returns {@code null} when it is not a
     * usable problem document (empty, not an object, or malformed JSON). Callers fall back to the
     * untyped {@code ApiException} in that case, so behaviour is never worse than the generated
     * default.
     */
    public static ParsedProblem parse(String body) {
        if (body == null || body.isBlank()) {
            return null;
        }
        try {
            JsonObject root;
            try (JsonReader reader = Json.createReader(new StringReader(body))) {
                JsonValue value = reader.readValue();
                if (value.getValueType() != JsonValue.ValueType.OBJECT) {
                    return null;
                }
                root = value.asJsonObject();
            }

            ProblemDetail problem = EpistolaJson.jsonb().fromJson(body, ProblemDetail.class);
            if (problem == null) {
                return null;
            }
            return new ParsedProblem(problem, readErrors(root), readValidationErrors(root));
        } catch (RuntimeException e) {
            return null;
        }
    }

    private static List<ValidationError> readErrors(JsonObject root) {
        JsonValue errors = root.get("errors");
        if (errors == null || errors.getValueType() != JsonValue.ValueType.ARRAY) {
            return Collections.emptyList();
        }
        return Arrays.asList(EpistolaJson.jsonb().fromJson(errors.toString(), ValidationError[].class));
    }

    private static Map<String, List<DataModelValidationError>> readValidationErrors(JsonObject root) {
        JsonValue validationErrors = root.get("validationErrors");
        if (validationErrors == null || validationErrors.getValueType() != JsonValue.ValueType.OBJECT) {
            return Collections.emptyMap();
        }
        Map<String, List<DataModelValidationError>> byExample = new LinkedHashMap<>();
        for (Map.Entry<String, JsonValue> entry : validationErrors.asJsonObject().entrySet()) {
            if (entry.getValue().getValueType() != JsonValue.ValueType.ARRAY) {
                continue;
            }
            JsonArray failures = entry.getValue().asJsonArray();
            byExample.put(
                    entry.getKey(),
                    new ArrayList<>(Arrays.asList(
                            EpistolaJson.jsonb().fromJson(failures.toString(), DataModelValidationError[].class))));
        }
        return byExample;
    }

    private ProblemDetailParser() {
    }
}

File: contracts/api/clients/nodejs-fetch/src/error/problemDetailParser.ts
// SPDX-FileCopyrightText: Epistola Nederland B.V.
//
// SPDX-License-Identifier: EUPL-1.2

import {
  DataModelValidationErrorFromJSON,
  ValidationErrorFromJSON,
  type DataModelValidationError,
  type ProblemDetail,
  type ValidationError,
} from '../generated/api/models/index.js'
import { ProblemExtensionMembers } from '../generated/knownProblemSlugs.js'
import { ProblemDetailException } from './problemDetailException.js'
import { BLANK_TYPE, PROBLEM_JSON } from './problemTypes.js'

const BASE_MEMBERS = new Set(['type', 'title', 'status', 'detail', 'instance'])

/** True when a `Content-Type` header value names the RFC 9457 problem media type, parameters aside. */
export function isProblemJson(contentType: string | null | undefined): boolean {
  if (!contentType) return false
  return contentType.split(';')[0].trim().toLowerCase() === PROBLEM_JSON
}

/**
 * Parses a problem+json `body` into a {@link ProblemDetailException}, or returns undefined on any
 * parse failure so the caller can fall back to the generic `ResponseError`.
 *
 * Parses the base {@link ProblemDetail} plus the field-level `errors` array
 * (`ValidationProblemDetail`) and the per-example `validationErrors` map
 * (`DataModelValidationProblemDetail`); the member names come from the contract. Every other
 * top-level member lands in `extensions`.
 */
export function parseProblem(body: string, statusCode: number, response: Response): ProblemDetailException | undefined {
  let tree: unknown
  try {
    tree = JSON.parse(body)
  } catch {
    return undefined
  }
  if (tree === null || typeof tree !== 'object' || Array.isArray(tree)) {
    return undefined
  }
  const node = tree as Record<string, unknown>

  const problem: ProblemDetail = {
    type: typeof node.type === 'string' && node.type !== '' ? node.type : BLANK_TYPE,
    title: typeof node.title === 'string' ? node.title : '',
    status: typeof node.status === 'number' ? node.status : statusCode,
    detail: typeof node.detail === 'string' ? node.detail : undefined,
    instance: typeof node.instance === 'string' ? node.instance : undefined,
  }

  let errors: ValidationError[] = []
  const rawErrors = node[ProblemExtensionMembers.ERRORS]
  if (Array.isArray(rawErrors)) {
    try {
      errors = rawErrors.filter(isRecord).map((entry) => ValidationErrorFromJSON(entry))
    } catch {
      errors = []
    }
  }

  const validationErrors: Record<string, DataModelValidationError[]> = {}
  const rawValidation = node[ProblemExtensionMembers.VALIDATION_ERRORS]
  if (isRecord(rawValidation)) {
    for (const [example, failures] of Object.entries(rawValidation)) {
      if (Array.isArray(failures)) {
        try {
          validationErrors[example] = failures.filter(isRecord).map((entry) => DataModelValidationErrorFromJSON(entry))
        } catch {
          // A malformed extension should not hide the problem it decorates.
        }
      }
    }
  }

  const extensions: Record<string, unknown> = {}
  for (const [key, value] of Object.entries(node)) {
    if (!BASE_MEMBERS.has(key)) {
      extensions[key] = value
    }
  }

  return new ProblemDetailException(problem, errors, validationErrors, extensions, statusCode, body, response)
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === 'object' && !Array.isArray(value)
}
